### PR TITLE
HPCC-14067 Avoid possible deadlock if spilling + callback called

### PR DESCRIPTION
**Pull request by jakesmith**
When the memory manager is actively trying to free memory at the
same time that a spilling stream is spilling and in the process
of deregistering it's callback a deadlock can occur - it can
block on a rows lock.
Avoid by checking if already spilt upfront in the callback and
exiting.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

File: thorlcr/thorutil/thmem.cpp [hunk 1]
```diff
@@ -214,6 +227,7 @@ class CSpillableStreamBase : public CSimpleInterface, implements roxiemem::IBuff
     	assertex(inRows.isFlushed());
```
Comment (richardkchapman) [inline thorlcr/thorutil/thmem.cpp]: Not new, but the indentation looks wrong on this line

**Comment by richardkchapman** [comment]
@mckellyln Can you check this too please

**Comment by mckellyln** [comment]
Is there a test case input I can use for this ?
Should mmRegistered be atomic ?

File: thorlcr/thorutil/thmem.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: it's a tab, there were a couple of other lines with tab's in the file too, have corrected.

**Comment by jakesmith** [comment]
@mckellyln - I have a test, but not one I could reproduce the problem with, without hacking the code to introduced both artificial delays + introducing a method into the MM to provoke the OOM callback mechanism at just the right time.

**Comment by jakesmith** [comment]
>Should mmRegistered be atomic ?

No, it should be protected by the mutex on row array (IThorArrayLock).
However, the additional clearSpillingCallback() in stop() wasn't, but now I've changed it so it is.

**Comment by jakesmith** [comment]
@richardkchapman @mckellyln - please re-review

File: thorlcr/thorutil/thmem.cpp [hunk 2]
```diff
@@ -392,7 +411,11 @@ class CSpillableStream : public CSpillableStreamBase, implements IRowStream
         ++pos;
         return row;
     }
-    virtual void stop() { }
+    virtual void stop()
+    {
+        CThorArrayLockBlock block(rows);
```
Comment (mckellyln) [inline thorlcr/thorutil/thmem.cpp]: Should we check spillFile here, before acquiring the lock and calling clearSpillingCallback() ?  Not sure, just asking.

File: thorlcr/thorutil/thmem.cpp [hunk 3]
```diff
@@ -201,9 +202,21 @@ class CSpillableStreamBase : public CSimpleInterface, implements roxiemem::IBuff
...
     }
+    void addSpillingCallback()
+    {
+        if (!mmRegistered)
+        {
+            mmRegistered = true;
+            activity.queryJob().queryRowManager()->addRowBuffer(this);
+        }
+    }
     void clearSpillingCallback()
```
Comment (mckellyln) [inline thorlcr/thorutil/thmem.cpp]: Could add and clear SpillingCallback be inlined ?

**Comment by jakesmith** [comment]
There were other paths to this deadlock, so I've changed the approach, ensuring that I don't call clearSpillingCallback, until outside the rows lock.
@mckellyln - please re-review

File: thorlcr/thorutil/thmem.cpp [hunk 4]
```diff
@@ -229,9 +243,30 @@ class CSpillableStreamBase : public CSimpleInterface, implements roxiemem::IBuff
...
+    inline CRowsLockBlock(CSpillableStreamBase &_owner) : owner(_owner)
+    {
+        owner.rows.lock();
+        clearCB = false;
+    }
+    inline ~CRowsLockBlock()
+    {
+        owner.rows.unlock();
+        if (clearCB)
+            owner.clearSpillingCallback();
```
Comment (mckellyln) [inline thorlcr/thorutil/thmem.cpp]: only question, should clearCB also be set to false here ?

Comment (richardkchapman) [inline thorlcr/thorutil/thmem.cpp]: Hardly matters in a destructor. It's about to cease to exist...

**Comment by mckellyln** [comment]
Otherwise looks good.

**Pull request merged**